pkg/workload/kv: add --zipfian-v and --zipfian-s flags

sean- opened the pull request:
* Update `workload kv --zipfian` to add `--zipfian-v` and `--zipfian-s` CLI flags to change rand.Zipf's data distribution

* Replace custom zipfian implementation with Go's built-in rand.Zipf

* Now initialize the Zipf RNG once, but implemented a custom rand.Source that iterates through a sequence of `--cycle-length`

Release note: Update `workload kv --zipfian` to add `--zipfian-v` and --zipfian-s` CLI flags to change the data distribution
Epic: none

sean- committed:
pkg/workload/kv: refactor kv Zipf RNG cycle

* Replace custom zipfian implementation with Go's built-in rand.Zipf

* Refactored sources don't immediately request a recently written value

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142258)
<!-- Reviewable:end -->

sean- committed:
workload: add zipfian S and V parameters to workload kv --zipfian

Epic: none

Release note: None

sean- committed:
fix linter

sean- committed:
workload: use crc64 as the zipfian hasher, not fnv64

Epic: none
Release note: None

sean- committed:
workload: update help for --zipfian and pick a random zipf value when reading

sean- committed:
add basic comments

BramGruneir reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142258#-:-OKbQpeU-10nOzNrIkLA:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DarrylWong and @srosenberg)



<!-- Sent from Reviewable.io -->

DarrylWong commented on pkg/workload/kv/kv.go:
```diff
@@ -942,40 +965,90 @@ func (g *sequentialGenerator) state() string {
+func (wss writeSequenceSource) Uint64() uint64 {
+	hasher := crc64.New(crc64.MakeTable(crc64.ECMA))
+	buf := []byte{0, 0, 0, 0, 0, 0, 0, 0}
+	// only increment the sequence during write calls
```
Super nit: perhaps this comment would be better in the `wss.write` method? Was confused thinking this function was doing the incrementing when it appears to be baked into `write`.

DarrylWong reviewed (approved):
LGTM. Mostly left understanding nits from the perspective of someone looking at `zipf` for the first time.

Would also be good kick off smoke test all roachtests that use `zipf` before merging if you haven't done so yet. Although surprisingly (to me) we actually don't run any `kvbench` with zipf distribution. It's just the `allocbench/*/ops=skew` tests it appears.

Edit: went ahead and kicked off a [run here](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/19024593?hideProblemsFromDependencies=false&hideTestsFromDependencies=false) since its just a couple of tests

srosenberg commented on pkg/workload/kv/kv.go:
```diff
@@ -942,40 +965,90 @@ func (g *sequentialGenerator) state() string {
+}
+
+func (wss writeSequenceSource) Uint64() uint64 {
+	hasher := crc64.New(crc64.MakeTable(crc64.ECMA))
```
`MakeTable` can get expensive on _every_ put. Why not reuse?

sean- replied:
`MakeTable()` caches the table creation and returns a value behind a `sync.Once`, so this call is actually cheap.

srosenberg replied:
🤦 I see, looked at the internal `makeTable`. In that case, you're still paying the price of alloc, but it looks like it should fit into the stack.

srosenberg commented on pkg/workload/kv/kv.go:
```diff
@@ -942,40 +965,90 @@ func (g *sequentialGenerator) state() string {
+}
+
+func (rss readSequenceSource) Uint64() uint64 {
+	hasher := crc64.New(crc64.MakeTable(crc64.ECMA))
```
Same as above. Consider reusing the `hasher`.

sean- replied:
Is `hasher` isn't reentrant and has state: https://cs.opensource.google/go/go/+/refs/tags/go1.24.1:src/hash/crc64/crc64.go;l=87-91

The above use of sha1 in this program is, I believe, incorrect for this very same reason, but I haven't confirmed.

srosenberg commented on pkg/workload/kv/kv.go:
```diff
@@ -942,40 +965,90 @@ func (g *sequentialGenerator) state() string {
+		sequence: seq,
+	}
+	return &zipfGenerator{
+		writeZipf: rand.NewZipf(rand.New(writeSS), zipfianS, zipfianV, uint64(math.MaxInt64)),
```
`rand/v2/NewZipf` takes advantage of more efficient `Float64`; why not v2?

sean- replied:
It's not directly API compatible with `math/rand` ... hold on, let me do that real quick.

srosenberg commented on pkg/workload/kv/kv.go:
```diff
@@ -942,40 +965,90 @@ func (g *sequentialGenerator) state() string {
+	// (don't increment the sequence).
+	v := rss.read()
+	if v != 0 {
+		// To prevent an immediate read of a recently written value, find a random
```
I'm not following what this is actually doing, and why we need a custom implementation of a uniformly-distributed source.

sean- replied:
We need a random value that's going to be on the Zipf distribution, which is what this does (i.e. we only want to read a valid value that's already been written to).

MattWhelan replied:
`rand.Zipf` is a utility that takes a `rand.Rand` (which produces a uniform random distribution) and produces a Zipfian distribution (approximately, one where the third element is 1/3 as likely as the first one). 

`rand.Rand` needs an entropy source, a `rand.Source`. The contract of `Source` is that it produces a uniformly distributed 63-bit (or 64-bit, for `Source64`) unsigned integer. 

Ultimately we want to be able to replay a deterministic sequence of events. We don't actually want random. We want a deterministic sequence that satisfies a distribution. So we start with a `sequence` which just iterates over an integer range and cycles, and for each value, we hash the value to spread them uniformly across the 64-bit integer space. That's our fake "entropy" source, which gives us the replayability that we want, and allows us to re-use the existing Rand and Zipf infrastructure.

sean- commented on pkg/workload/kv/kv.go:
```diff
@@ -99,6 +103,21 @@ func init() {
+	}
+)
+
+func randSeedGet() int64 {
```
I didn't want to risk it being overwritten by a CLI arg vs env var.

sean- committed:
add comments

sean- committed:
pkg/workload: bump to use math/v2/rand

sean- committed:
pkg/workload: improve zipfian-related docs

sean- commented on pkg/workload/kv/kv.go:
```diff
@@ -151,13 +170,17 @@ var kvMeta = workload.Meta{
 		g.flags.BoolVar(&g.zipfian, `zipfian`, false,
-			`Pick keys in a zipfian distribution instead of randomly.`)
+			`Allocate keys using a zipfian distribution instead of a uniform distribution of random keys.`)
+		g.flags.Float64Var(&g.zipfianS, `zipfian-s`, 1.1,
```
See comments and updated CLI flags

sean- committed:
pkg/worload/cli: fix --ramp race condition

Periodically `workload run --ramp` would hang.  Moving rampDone out of the for
loop also reduced CPU utilization by ~50% in my testing.  Improve readability
by refactoring warmup by applying modern Go improvements via context.Context
and sync.WaitGroup.

Epic: none
Release note: None

sean- commented:
bors r+

sean- commented:
bors r+
